### DISCUSSION TITLE
Product Gallery block: Core gallery solution limitations and their implications for development

### DISCUSSION THREAD

**Discussion by danieldudzic** [woocommerce/woocommerce#44966] [WooCommerce, Blocks, and Site Editing]
### Background

Recently @thealexandrelara and I started working on a [successor of the existing **Product Image Gallery** block](https://github.com/woocommerce/woocommerce-blocks/issues/9891#event-9572382102). 

The project aims to provide enhanced **Product Gallery** customizability in the context of the recently blockified **Single Product** template (and in the future also [other blocks like the Single Product](https://github.com/woocommerce/woocommerce-blocks/issues/10020)).

| Structure | Features |
| ------ | ----- |
|![cleanshot-2023-05-28-at-11 11 07402x](https://github.com/woocommerce/woocommerce-blocks/assets/905781/ba70caf6-70d4-4423-829b-6d9a39eb874f)| <img width="2549" alt="cleanshot-2023-05-22-at-12 09 01402x" src="https://github.com/woocommerce/woocommerce-blocks/assets/905781/90af10ad-3aad-4ffd-9b99-6dbb3e731eeb">|

### Current state of the Product Gallery

The current Product Gallery lives in the following template files:
- [photoswipe.php](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/single-product/photoswipe.php)
- [product-image.php](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/single-product/product-image.php)
- [product-thumbnails.php](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/single-product/product-thumbnails.php)
- [sale-flash.php](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/single-product/sale-flash.php)

and utilizes:
- [Flexslider](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js) - handles behavior of the gallery: animations, "slide" changes.
- [Photoswipe](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/js/photoswipe/photoswipe.js) - handles the Lightbox experience (after the Magnifying glass icon on the Large Image is selected).
- [Zoom](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/js/zoom/jquery.zoom.js) - handles the zoom effect on hover on the Large Image.

### Current state of the Product Image Gallery block

It is essentially a [product-image.php template wrapper](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTypes/ProductImageGallery.php).

### Problem statement - Limitations in utilizing the existing solution vs new blockified experience goals

We have discussed whether we should keep the existing **Product Image Gallery** block as a classic _back-compat_ block and introduce a new **Product Gallery** block or try to integrate the new experience into the existing block. 

Even though we have given ourselves the freedom to create a new separate block, we have organically ended up trying to see if we can leverage the existing Woo core Product Gallery solution, and expand on utilizing existing templates and actions.

This approach, of course, isn't very innovative and cutting-edge but we are hoping we can tap into the existing solution and quickly deliver a working block that provides a lot of customization options and doesn't break compatibility with the existing Core hooks and filters.

An example of this is the [**Large Image** block](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTypes/ProductGalleryLargeImage.php).

It does pretty much the same thing as the **Product Image Gallery** but just [removes the thumbnails](https://github.com/woocommerce/woocommerce-blocks/blob/d59afbce01e374d207d4bf47292ba97d2888d4a3/src/BlockTypes/ProductGalleryLargeImage.php#L66).

As we dig further into separating elements of the gallery into separate blocks we start encountering problems with this strategy.

When working on [**Next/Previous buttons** block](https://github.com/woocommerce/woocommerce-blocks/issues/9901) I have discovered that even though you can enable those in Flexslider via:
```
		add_filter(
			'woocommerce_single_product_carousel_options',
			function( $options ) {
				$options['directionNav'] = true;
				return $options;
			}
		);
```

This applies to all Flexsliders on the page and it is still connected to the Flexslider (Large Image) gallery markup.

The workaround here is to recreate the Next/Previous buttons markup and then utilize the Flexslider API: `flexsliderInstance.flexslider( 'prev' );` and `flexsliderInstance.flexslider( 'next' );` to add desired functionality.

There's an additional caveat, this only works if the thumbnails are not being removed via:
`remove_action( 'woocommerce_product_thumbnails', 'woocommerce_show_product_thumbnails', 20 );` (which is what the **Large Image** block does by default). If the thumbnails are not present, the markup for other images (other than the main one) is simply not being loaded, and hence there's nothing to slide/navigate to with the buttons.

I feel we could overcome this by first loading a bunch of stuff and then just filtering out selectively via regex, but this is starting to feel very hacky.

### Proposed solutions

I can think of the following ways to continue the development of the **Product Gallery** block.

### 1. Continue hacking away at it

We can try and manipulate the HTML and styling as much as possible to keep leveraging the current **Product Gallery** scripts and styling.

**Pros**: It's quick. We maintain the current Core familiar UX. We maintain compatibility* (assuming we don't break it with all that _filtering_).
**Cons**: It's hacky. The Core solution is _a bit_ dated.

### 2. Load custom templates, but leverage current scripts and styling

We can load our own "templates" in a way that will be compatible with the existing Core styling and gallery scripts (leverage the Flexslider API for example).

**Pros**: It's less hacky. We maintain the current Core familiar UX. We can try and keep it hook and filter compatible.
**Cons**: The Core solution is _a bit_ dated. It will be more time-consuming to develop. We may still run into issues with the limitation of the API and scripts.

### 3. Mix and match

We load custom "templates", but we keep *some* markup in a way that will be compatible with the existing Core styling and gallery scripts. For example, we decide to keep **photoswipe**, but we use a new, custom slider script for the gallery.

**Pros**: It's even less hacky. We maintain some of the Core familiar UX. We can try and keep it hook and filter compatible.
**Cons**: The Core solution is _a bit_ dated. It will be even more time-consuming to develop.

### 4. Out with the old, in with the new

We ditch the Core holdover experience altogether. We introduce new "templates", scripts, and styling.

**Pros**: It's clean and elevates the UX. 
**Cons**: It will require a significant time to develop. It breaks compatibility with Core hooks and filters. It's unfamiliar to users.

### Suggested next steps

I think I'm leaning towards **option no. 2**. I know it's not perfect, but if we can have a working block, relatively quick, it's a win. I think the aforementioned pros of that solution are somewhat convincing. I do fear a little bit, that we will run into the same problems as with the current (no. 1) strategy.

I'm also tempted to continue giving the **option no. 1** a quick try to see how far we can get, the dirty way.

### Feedback request

I’d appreciate your comments and ideas regarding the best path forward in **Product Gallery** block development 🙏

Please feel free to correct misunderstandings or invalid assumptions I might have made in the above analysis.

CC: @thealexandrelara, @tjcafferkey, @danielwrobert, @nerrad

**Comment by webaxones** [comment]
Greetings from an external developer,

If my message does not bring you anything, ignore or delete it, but just to give you a glimpse from our side: here are the hooks that I use on the product gallery on almost all my ecommerce sites.

```php
/**
 * Remove zoom from single product page lightbox
 *
 * @return void
 */
function remove_zoom_lightbox_theme_support(): void {
	remove_theme_support( 'wc-product-gallery-zoom' );
}
add_action( 'wp', __NAMESPACE__ . '\remove_zoom_lightbox_theme_support', 99 );

/**
 * Customize single product photoswipe lightbox options
 *
 * @param array $array Photoswipe options.
 *
 * @return array Photoswipe options.
 */
function woocommerce_single_product_photoswipe_options_filter( $array ): array {
	$array['closeOnScroll']         = true;
	$array['hideAnimationDuration'] = true;
	return $array;
}
add_filter( 'woocommerce_single_product_photoswipe_options', __NAMESPACE__ . '\woocommerce_single_product_photoswipe_options_filter' );

/**
 * Customize single product gallery thumbnail size
 *
 * @param array $size Gallery thumbnail size.
 *
 * @return array Gallery thumbnail size.
 */
function customize_image_size_gallery_thumbnail( $size ): array {
	return [
		'width'  => 300,
		'height' => 165,
		'crop'   => 0,
	];
}
add_filter( 'woocommerce_get_image_size_gallery_thumbnail', __NAMESPACE__ . '\customize_image_size_gallery_thumbnail' );

```
(This example comes directly from an ongoing redesign in FSE mode with WC Blocks and use of the Product Gallery block.)

**Reply by nerrad** [reply]
Thanks for jumping in @webaxones - your insight is very valuable and I appreciate you taking the time to leave it here. Given the opportunity we have here to improve not only the end-user experience for the block (UX), but also the developer experience in extending it or defining defaults (DX), if some of the options you mentioned were configurable at the `theme.json` level (including the ability to lock down changes), would that be helpful?

**Reply by webaxones** [reply]
Hi Nerrad,

> if some of the options you mentioned were configurable at the theme.json level (including the ability to lock down changes), would that be helpful?

Of course!
Hook or theme.json, the result is the same for us.
Now, we can question the capacity of the growing theme.json. It becomes a large file to load into memory, requiring CPU time to decode/encode the JSON in memory... but that's not specific to WC.

**Comment by nerrad** [comment]
Something we'll want to keep an eye on in WordPress is the [Block Behaviours API](https://github.com/WordPress/gutenberg/issues/50029) that is likely something that Woo would integrate with in blocks like the Featured Image Gallery. It'd be good to explore if there are design patterns being implemented there that we could experiment with sooner in Woo Blocks to find the boundaries of that API and how well it meets the needs of this block.

**Reply by danieldudzic** [reply]
Thank you for sharing. I will give the Block Behaviours API thread a thorough read.

**Comment by tjcafferkey** [comment]
Thanks for this Danny. Here are my thoughts, but I am totally fine with leaning on yourself and @thealexandrelara to make the decision here as you are closer to the implementation details than I am. I also appreciate how you've outlined multiple options and their pros/cons in tackling this problem.

Ideally, we do not want to rely on old resources as dependencies to our new features especially if those old resources have either stopped being maintained or are not really aligned with the technical direction we're taking with new features (e.g. uses jQuery).

However, given that we don't always have time on our side it sounds like (based on what you have said above) that Option 2 is the most feasible for us to have something to ship at the end of our current development cycle, is that right? Something I'd also like to know is if we _do_ decide to go with Option 2, is that we have the ability to iterate on this block and replace legacy scripts with our own functionality?

So I guess what I'm asking is if our end goal is to have a block that doesn't depend on underlying legacy code, is Option 2 still the one you would go with, and does it take us a step forward in that direction?

**Reply by thealexandrelara** [reply]
> Ideally, we do not want to rely on old resources as dependencies to our new features especially if those old resources have either stopped being maintained or are not really aligned with the technical direction we're taking with new features (e.g. uses jQuery).

I agree with you; we should avoid relying on old resources. The current implementation of the Product Image Gallery utilizes Photoswiper, Zoom, and Flexslider, all of which appear to be part of legacy code (based on the folder they are located in).

> However, given that we don't always have time on our side it sounds like (based on what you have said above) that Option 2 is the most feasible for us to have something to ship at the end of our current development cycle, is that right?

Yes, that is the idea. We are trying to speed up things by reusing as much of the current implementation as we can. However, we are still exploring the limitations because even if we manage to reuse most of the current implementation and split it into several inner blocks, there are many improvements that we want to make to this block. We also want to ensure it is prepared for future enhancements.

> Something I'd also like to know is if we do decide to go with Option 2, is that we have the ability to iterate on this block and replace legacy scripts with our own functionality?
>
> So I guess what I'm asking is if our end goal is to have a block that doesn't depend on underlying legacy code, is Option 2 still the one you would go with, and does it take us a step forward in that direction?

I think this is something we need to explore, but with Option 2 I'm not sure how much flexibility we would encounter to avoid using the legacy scripts. 

Since the ideal scenario is to avoid using legacy scripts and we have three legacy scripts being used, maybe we can try replacing one of them, and see if we can make everything work without breaking the other parts, once we succeed with that we can proceed with replacing the others (this approach would be closer to the option 3). This way we can measure how much legacy code can be eliminated and replaced with newer alternatives.

**Comment by nerrad** [comment]
I appreciate the tension between completing something this cycle and arriving at the optimal solution for the end user as represented in the options you've presented. To expand on what Tom has already suggested, I think there is also an opportunity to remove reliance on large dependencies such as jQuery (via flexslider) and look for alternatives that could provide similar (and eventually enhanced) functionality for users of the block _that also can be integrated with the new [Behaviours API](https://github.com/WordPress/gutenberg/issues/50029)_ coming to WordPress (as I mentioned earlier). For instance, there's already some established explorations around a Lightbox behaviour (as a part of the [initial Interactivity API implementation](https://github.com/WordPress/gutenberg/issues/52292)) that might eventually be re-usable for our needs. In the interim we can likely replicate that functionality given we have a custom implementation of the Interactivity API in Woo Blocks. 

Have you looked into any alternatives for replacing the jQuery flexslider plugin?

Another important consideration in the refinement of this block is what was raised by @webaxones in this thread that indicates there are various ways developers have customized the behaviours on the frontend for this block. With the interactivity API, there _may_ be some server side customization options but ideally, by using the Behaviours API there will be some opportunities for extensibility to be built in to that API so feasibly developers can enhance any blocks with their own custom behaviour.

Before making a decision on the _existing_ options outlined in this discussion, I do think the above needs some more investigation.

**Reply by thealexandrelara** [reply]
Thank you, @nerrad, for bringing up these points.

I completely agree that we should try to reduce our reliance on the large dependencies currently in use and look for alternatives. I haven't heard of the Behaviours API before, but it sounds awesome. I believe it can provide the flexibility and customizability we are aiming for with this block. Therefore, it's something we can explore in terms of preparing for this new API and integrating it with the block.

Considering the points you raised, before we make any decisions on the options, we should:

1. Find alternatives to the large dependencies we currently rely on (that can be used together with the Behaviours API).
2. Attempt to replicate the Lightbox behavior in WooCommerce Blocks project (using the Interactivity API).
3. Explore ways to allow developers to customize the behavior of this block (using the Interactivity API and/or Behaviours API).

**Comment by gigitux** [comment]
> I appreciate the tension between completing something this cycle and arriving at the optimal solution for the end user as represented in the options you've presented. To expand on what Tom has already suggested, I think there is also an opportunity to remove reliance on large dependencies such as jQuery (via flexslider) and look for alternatives that could provide similar (and eventually enhanced) functionality for users of the block that also can be integrated with the new https://github.com/WordPress/gutenberg/issues/50029 coming to WordPress (as I mentioned earlier). For instance, there's already some established explorations around a Lightbox behaviour (as a part of the https://github.com/WordPress/gutenberg/issues/52292) that might eventually be re-usable for our needs. In the interim we can likely replicate that functionality given we have a custom implementation of the Interactivity API in Woo Blocks.


I agree with this approach. Today we had a sync call about the Product Gallery block and discussed a roadmap that allow us to move on and simultaneously explore different approaches.

The goals that we are defined are:
- Compatibility with Single Product Block and Single Product Template.
- Elimination of jQuery and outdated technologies.
- Implementation of Interactivity API and Behavior API.
- Compatibility Layer (if we decide to get rid of the current Product Gallery and migrate the user to the new one and/or we can't reach the right level of extendibility via the Beavhiours API)
- E2E tests.

To achieve these goals, we have outlined a roadmap as follows:

- Lay the foundation of the block, both on the editor and frontend sides.
    - Develop the necessary settings for the block.
    - Design the user interface (UI) for the block.
    - Write the markup on the editor side and front end side.
    - Implement E2E tests to ensure robustness.
- Integrate the Interactivity API.
- Implementing the Behavior API to extend the block's functionality.
- Creating the Compatibility Layer (if needed).
- Add E2E tests

Our initial focus will be on the markup and settings for both the editor and frontend. This approach allows us to identify and gather all the filters and actions that will no longer be available. By establishing a solid foundation, we can proceed to integrate the Interactivity API, which will introduce dynamicity to the block. Simultaneously, we will explore the behavior API to ensure that we can reach a good level of customization.

What do you think?

cc @danielwrobert @danieldudzic @thealexandrelara

**Reply by danieldudzic** [reply]
Thank you @gigitux for summarizing the call.

I agree with the steps outlined above. Focusing on block foundations and adding settings seems like a solid, achievable (first) step forward. 

After weighing all the options, and discussing it on the call I now believe that building the new Product Gallery from scratch will ultimately deliver best results, even if it will take a little longer.
